### ISSUE TITLE
Silent miscompilation of masked assignment, plus 3 defects and a bf16 regression, on a dilated CNN (trn2, neuronxcc-2.26.6360.0)

### ISSUE THREAD

**Issue by cralbors** [aws-neuron/aws-neuron-sdk#1381]
# neuronx-cc defects found training a dilated CNN on trn2.48xlarge

Five findings from porting a dilated CNN to Trainium2. One is a **silent
miscompilation that returns wrong values with no error raised** — that one matters
most.

## Environment

```
instance    trn2.48xlarge (16 chips, 64 logical NeuronCores at LNC=2)
AMI         Deep Learning AMI Neuron (Ubuntu 24.04)
torch       2.9, torch-neuronx 2.9.0.2.15.32035+de43f57c
compiler    neuronxcc-2.26.6360.0+6f180f47
libneuronxla 2.2.17544.0+fb9962bf
```

Model shape where these appear: `model_dim=256`, `num_stacks=6`,
`kernel_sizes=[11,5,5,5,5,5]`, dilations 1..32, `feedforward_dim=2048`,
sequence length 5548, fp32.

---

## 1. SILENT MISCOMPILATION — boolean-mask assignment returns wrong values

**Severity: high.** No error, no warning, finite output, wrong numbers.

A masked in-place assignment whose selection has a data-dependent shape:

```python
expanded = stored[expansion_indices]                       # (N, K)
expanded[needs_flip_mask] = expanded[needs_flip_mask].flip(-1)
```

`expanded[mask]` has a shape that depends on the mask's contents, so it lowers to
a dynamic gather/scatter. On CPU this is exact. On Trainium it compiles, runs,
and produces **wrong values**.

Measured: a single `Gated1DConvolution` built with such weights had forward
cosine **0.856** against CPU (relative error 0.92), compounding to ~0.49 over 72
blocks. Nothing was raised at any point.

Replacing it with a static gather — an index built for every row, flipped or not —
made the full 72-block forward **bit-exact** against CPU (`max_abs = 0.000e+00`):

```python
kernel_index = torch.where(mask.unsqueeze(-1), reverse_index, forward_index)
expanded = expanded.gather(-1, kernel_index)
```

We only found this because we ran a numerical parity check. A user who trusts
"it compiled and the loss is finite" would train a corrupted model indefinitely.

---

## 2. NCC_INLA001 — `torch.flip` feeding a matmul cannot be lowered

```
[INTERNAL_ERROR] [NCC_INLA001] === BIR verification failed ===
Reason: RHS AP cannot have negative stride
Instruction: I-19383   Opcode: Matmult   Input index: 1
Access Pattern: [[128,1],[-1,128]]
Memory Location: {reduce.2_pftranspose_7669_set}@SB0,0
```

A reverse lowers to a negative stride, and a matmul whose right-hand operand is
read that way fails the compiler's own IR verifier. Reproduced by a single
`FeedforwardNetwork` at width 5548 in 30 s.

Fails identically at `--optlevel=1`, `--model-type=cnn-training` and
`--auto-cast=none`. At `--optlevel=3` a different lowering avoids it and reaches
the instruction ceiling instead.

---

## 3. NCC_IBIR243 — the obvious workaround wraps -1 into an unsigned value

Rewriting the same reverse as `index_select` with descending indices:

```
[INTERNAL_ERROR] [NCC_IBIR243] Access pattern out of bounds.
Pattern: [[4294967295,128],[1,1],[1,1]]
```

`4294967295` is `-1` as unsigned 32-bit. The compiler folds the gather back into
a reverse and the negative stride reappears, this time as an out-of-bounds
unsigned access rather than a caught error. Worse than #2 because the failure
mode is a wraparound rather than a diagnostic.

---

## 4. bf16 is 6.6-8x SLOWER than fp32

Same model, same shape, one training step (median of 30, single core):

```
fp32                                    9.894 s
NEURON_CC_FLAGS=--auto-cast=all bf16   65.252 s   6.6x slower
model.to(torch.bfloat16)               79.265 s   8.0x slower
```

Both routes regress, which suggests convert-op proliferation or missing bf16
kernels for dilated conv / gather / LayerNorm rather than one bad kernel.

Note `--auto-cast=none` produced **bit-identical** gradients to the default, so
the default is not casting matmuls either.

This is the largest single performance surprise: bf16 is the hardware's headline
capability and it is unusable here.

---

## 5. Question, not a defect — SBUF tiling for long sequences

Batch is capped at 2 by the state buffer, not by HBM:

```
[NCC_IBIR229] State buffer allocation failed:
  multiply.13_sub0:      22,806,528 bytes in SB
  multiply.13_2884_sub0: 22,986,752 bytes in SB
  Total available SB Size: 29,360,128
```

One activation at batch 4 is exactly `4 x 5548 x 256 x 4 = 22,724,608` bytes, and
an op needs two resident against 28 MiB. HBM is 24 GB and nearly empty.

Is tiling along the sequence dimension possible for these ops? A GPU never has to
answer this because caching is implicit. For long-sequence CNNs — common in
genomics — this single constraint costs an order of magnitude: an A100 runs batch
24 on the same model, we run batch 2.

Related: a 5,000,000-instruction ceiling means a 72-layer backward
(31,125,638 instructions) must be hand-split into segments with `mark_step`.
Workable, but it multiplies compile count and wall-clock.

---

## Reproducers

Self-contained scripts, each failing in under a minute on one core:

- `probe_bisect.sh` — #2, one module at width 5548
- `probe_conv.sh` — #1, isolates the parametrized conv against stock `nn.Conv1d`
- `probe_perf.sh` — #4, fp32 vs both bf16 routes
- `probe_batch2000.sh` — #5, the SBUF ceiling as batch climbs

Happy to attach or reduce further on request.

**Comment by cralbors** [comment]
## 6. `NCC_IBVF038` — bitvec `TensorReduce` rejects `int8`, the dtype XLA uses for `pred`

A sixth defect, found after filing, on the same environment as the parent issue.

`torch.logaddexp` fails to compile:

```
(TensorReduce: I-9585), tensorizer(
  output tensor: int8<128 x 34> $9585, id: 6126
  output tensor: int8<128 x 2>  $9586, id: 6126
  output tensor: int8<104 x 34> $9589, id: 6126 ...)
[INTERNAL_ERROR] [NCC_IBVF038] Non-C bitvec TensorReduce requires dtype in
UINT8, UINT16, UINT32, INT32, got in=int8 out=int8
```

Source line:

```python
log_y = torch.logaddexp(log_y, log_a.unsqueeze(-1) + log_v)
```

### Why this looks like a compiler-side dtype gap

`logaddexp` lowers to a max-plus-`log1p` form whose comparison emits an XLA `pred`. XLA represents `pred` as **int8**. The bitvec `TensorReduce` accepts `UINT8, UINT16, UINT32, INT32` — every integer width except the signed 8-bit one the frontend actually produces for predicates.

So a stock PyTorch op on float32 inputs cannot reach a supported dtype from user code: nothing in the Python source names int8, and there is no cast that survives the lowering.

### Reachable on most but not all shapes

```
batch 1, chunk 4000     NCC_IBVF038
batch 1, chunk 2000     NCC_IBVF038
batch 2, chunk 2000     NCC_IBVF038  (130 occurrences, killed a 32-rank run)
batch 2, chunk 4000     compiles     (fails later, on device HBM)
```

So it is not purely a dtype-dispatch gap: one shape in the grid selects a lowering that avoids the int8 reduce entirely. We cannot tell from outside which property of that shape is responsible.

What makes this hard to work around is exactly that — the same unchanged source compiles or does not depending on dimensions chosen for unrelated reasons, and the one shape that avoids it here is ruled out by a different limit.

### Workaround

Replacing `torch.logaddexp` with the max-shift identity compiles:

```python
SHIFT_FLOOR = -1e30                       # finite, below any real log-prob
shift = torch.maximum(a, b).clamp(min=SHIFT_FLOOR)
result = shift + torch.log(torch.exp(a - shift) + torch.exp(b - shift))
```

Note the `clamp` rather than the more natural `torch.where(torch.isinf(m) & (m < 0), 0, m)` guard for the all-`-inf` case: that guard builds boolean masks, which are the int8 predicates that trigger this defect in the first place. Pure arithmetic avoids producing the dtype at all.

Verified against stock `logaddexp` on CPU — max abs error 1.4e-14 in float64 and 3.8e-6 in float32, gradients matching to 2.2e-16, and identical `-inf` handling with no `nan`.

### Impact

This is on the loss path, not the model. Any model using `logaddexp` on a path the compiler fuses into a reduce will hit it, and the diagnostic points at a dtype the user never wrote and cannot cast away from the frontend.

### Environment

Same as the parent issue: trn2.48xlarge, `neuronxcc-2.26.6360.0+6f180f47`, `torch-neuronx 2.9.0.2.15.32035+de43f57c`, fp32.

**Comment by cralbors** [comment]
## Correction: the bf16 slowdown does not reproduce at operation granularity

The original report states that bf16 is 6.6–8× slower than fp32 on this model. We have not been able to reproduce that at the level of individual operations, and we want to correct the record rather than leave a number standing that we can no longer support.

### What we measured

Eight operations at the model's real shapes — `conv1d` at dilation 1 and 32, the gated conv block (fused filter+gate, SiLU, multiply), a 256×256 matmul, the 256→2048→256 feedforward, layernorm, and SiLU — in both the forward and the forward+backward direction, fp32 versus bf16. Width 2000, model_dim 256, ff_dim 2048, kernel 11.

```
                    fp32       bf16     ratio
FORWARD
conv1d d=1        73.9ms     81.8ms     1.11
conv1d d=32       76.8ms     78.9ms     1.03
gated conv d=1    93.7ms    101.8ms     1.09
gated conv d=32   97.1ms     98.5ms     1.01
matmul 256x256    36.8ms     38.3ms     1.04
feedforward       72.3ms     79.3ms     1.10
layernorm         49.0ms     48.4ms     0.99
silu              20.9ms     20.2ms     0.97
FWD+BWD
conv1d d=1       174.1ms    184.2ms     1.06
conv1d d=32      183.3ms    185.0ms     1.01
gated conv d=1   568.7ms    285.8ms     0.50   <- see caveat
gated conv d=32  275.7ms    296.6ms     1.08
matmul 256x256    59.7ms     60.7ms     1.02
feedforward      235.3ms    229.2ms     0.97
layernorm         97.8ms     97.2ms     0.99
silu              45.1ms     43.3ms     0.96
```

bf16 costs roughly 5% more than fp32 in the forward pass and is at parity in the backward. Nothing here resembles a 6.6–8× regression.

### A methodology note that may be useful to others

Our first attempt at this measurement timed one operation per graph and returned a flat 360–390 ms for **every** operation, regardless of dtype, dilation, or how much arithmetic the operation contained. A gated conv — double the convolution work plus a SiLU and a multiply — came in 8% above a plain conv, and dilation 1 and dilation 32 were indistinguishable.

That was per-graph dispatch cost, not compute. Any real difference was buried underneath a constant far larger than the thing being measured, and the ratios taken from it scattered from 0.67 to 1.40 with no pattern.

The numbers above come from chaining each operation into itself 24 times inside a single graph, so one synchronization covers 24 operations and the fixed cost amortizes to ~1/24. Chained rather than summed, because independent identical copies are common-subexpression-eliminated down to one — a real data dependency is required to keep them.

### Caveat

The `gated conv d=1` backward cell is an outlier we do not trust. bf16 is steady at 286–297 ms across both dilations while fp32 reads 568.7 ms at d=1 and 275.7 ms at d=32 — twice as slow at the cheaper dilation, which is not physically sensible. Three repeats per cell is evidently not enough to stabilize timings on this stack. We would treat any single cell here as ±2× until re-measured with more repeats and a reported variance.

### What we are not withdrawing

The whole-model observation was measured, on both the compiler-autocast path and explicit casting. We are only saying it does not localize to any individual operation, which means the cause is more likely graph count, layout transitions between operations, or collectives than any operation being slow in bf16. We will follow up if we manage to localize it.

---

## Amendment to finding 1: the miscompilation is shape-dependent

Finding 1 reported `expanded[mask] = expanded[mask].flip(-1)` producing cosine 0.856 against CPU at production dimensions.

At shape `(64, 64, 11)` the same source is **bit-exact**. Anyone attempting to reproduce at a convenient small size will conclude it does not reproduce. The production shapes given in the original report are required.

**Comment by cralbors** [comment]
**Correction to the above.** Further testing on the same environment contradicts
three claims in my previous comment. Posting the retraction rather than leaving
it to stand, since the attribution there could send someone down the wrong path.

**1. `logaddexp` is probably not the trigger.** I isolated the pieces of the
CTMC loss on one core and they all compile, forward and backward, at the
production width:

```
logsumexp + logaddexp truncation loop (12 iterations)   PASS
uniformize   (logsumexp, max(-1), eye mask, -inf test)  PASS
decompose    (masked_fill, min(-1), boolean OR)         PASS
full log_transition_matrix_mvp composition              PASS
where(mask, X, 0.0).sum(-1)                             PASS
```

**2. The max-shift workaround is not sufficient.** It does clear the forward,
and the numbers quoted for it are accurate. But the same `NCC_IBVF038` then
appears in the *backward*, 126 occurrences. Note that `torch.maximum` and
`clamp` build comparison masks in their own backward, so an attached shift moves
the defect rather than removing it; `.detach()` on the shift removes those (the
gradient is unchanged, since the shift cancels analytically). Even with that,
the defect persists.

**3. It is not confined to the loss path.** With one backbone block per segment,
the failure lands in a backbone segment's backward, not in the loss.

**Where it actually appears.** No individual module reproduces it. A single
`Gated1DConvolution` and a single `FeedforwardNetwork` each compile forward and
backward at width 4000. It only appears once the graphs are composed, which is
what makes it hard to reduce further from outside the compiler.

**Possibly the most useful detail for you:** the int8 tile shapes are identical
across every occurrence and across two different sequence lengths —
`int8<128 x 34>`, `int8<128 x 2>`, `int8<80 x 34>`, `int8<104 x 34>`. That looks
like a tiling artifact rather than anything derived from our tensor dimensions.

**A lead we cannot confirm, offered because it may matter more than the rest.**
Training now completes steps on this model for the first time, after setting:

```
XLA_PARAMETER_WRAPPING_THREADSHOLD=1000000      # torch-xla default is 3200
```

Our model presents far more than 3200 parameters to the graph, because an
RC-equivariant parametrization rebuilds 435 weight tensors in-graph on every
forward, on top of the real parameters. Past the threshold torch-xla wraps them
into a tuple, and with the wrapped form the run failed; unwrapped, it steps.

We are explicitly **not** claiming this causes `NCC_IBVF038` — we did not test
the same shape both ways, so we have correlation and a plausible mechanism, no
more. But if models that cross that threshold take a different lowering path,
that would be worth checking on your side, and it may explain why the failure
only appeared in composed graphs and never in any isolated module we probed.

(Noting also that the variable is misspelled in torch-xla itself —
`THREADSHOLD`, not `THRESHOLD`. Spelled correctly it has no effect, which cost
us time.)

Apologies for the noise on the first version.

**Comment by aws-yishanm** [comment]
Hi @cralbors, could you share the 4 reproducer scripts with us?
